File: app/src/main/java/com/example/mvvm_artichecture_sample/base/network/base/BaseRetrofitClient.java
package com.example.mvvm_artichecture_sample.base.network.base;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public abstract class BaseRetrofitClient {

    public <T> T createRequest(final Class<T> tClass) {

        final Retrofit retrofit = new Retrofit.Builder().baseUrl(getBaseUrl())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .client(getClient())
                .build();
        return retrofit.create(tClass);
    }

    protected abstract OkHttpClient getClient();

    protected abstract String getBaseUrl();

}

File: app/src/main/java/com/example/mvvm_artichecture_sample/base/network/base/BaseRetrofitClient.kt
package com.example.mvvm_artichecture_sample.base.network.base

import com.jakewharton.retrofit2.adapter.kotlin.coroutines.CoroutineCallAdapterFactory
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

abstract class BaseRetrofitClient {

    protected abstract val client: OkHttpClient

    protected abstract val baseUrl: String

    fun <T> createRequest(tClass: Class<T>): T {
        val retrofit = Retrofit.Builder().baseUrl(baseUrl)
                .addCallAdapterFactory(CoroutineCallAdapterFactory())
                .addConverterFactory(MoshiConverterFactory.create())
                .client(client)
                .build()
        return retrofit.create(tClass)
    }

}
